### ISSUE TITLE
REST API - Dataset Endpoint

### ISSUE THREAD

**Issue by jkglasbrenner** [usnistgov/dioptra#17]
## User Story

As a user or administrator of the Tested, I would like to be able to use an endpoint to register what the available datasets are on the storage backend and indicate what kinds of use cases or experiments they can be used for. This is so that I can more easily swap the datasets in and out as another variable in my experiments.

## Implementation

This user story creates a set of endpoints under a `/api/dataset/` namespace for registering and managing the datasets within a deployment.

### Dataset Structure Specification

To keep the registration details simple, the datasets will need to be standardized before they can be accepted into the API's database. This means that we need to create a set of specifications for how to structure (meaning folder structure) each kind of dataset before we can start implementing the endpoints themselves. For example, image data for classification problems might follow this kind of folder structure,

```
./name-of-image-dataset
├── testing-full
│   ├── label0
│   ├── label1
│   ├── label2
│   └── label3
├── testing-small
│   ├── label0
│   ├── label1
│   ├── label2
│   └── label3
├── training-full
│   ├── label0
│   ├── label1
│   ├── label2
│   └── label3
└── training-small
    ├── label0
    ├── label1
    ├── label2
    └── label3
```

Here, the image dataset is divided into four subgroups and the image files are stored in the `label0`, `label1`, `label2`, and `label3` folders. 

### Required Capabilities

These endpoints, at a minimum, need to allow a user to do the following:

- While image data is the current use case, the endpoints need to be flexible enough so that it can be easily extended to register additional kinds of data, such as unstructured text data, audio data, and tabular data.
- While classification of data with discrete target labels is the current use case, the endpoints should be flexible enough that they can be extended to handle data with continuous targets (regression).
- Users should be able to query a dataset registration record using a unique ID or a unique name
- Users should be able to query where the registered dataset is stored, i.e. the root directory on an NFS mount, the `s3://` URI in Minio, etc., so that it can be used to point at datasets within the task plugins
- Retrieve information about the dataset subgroups based on the registered dataset's folder structure
- Retrieve summary information about the dataset itself. For image data, this may include the total number of images in the dataset, the number of labels, the number of images per label, distribution of image dimensions, number of color channels, and so on.
- Allow users to add unstructured descriptions to the dataset
- Allow users to understand the kinds of models and machine learning problems the dataset is compatible with
- Allow users to generate random subsamples of a dataset to make it easier to test out new task plugins. This requires providing enough information that a task plugin could use soft linking or selective copying to generate the dataset subsample
- For image datasets, users should be able to download individual images (later on, a frontend would allow these images to be viewed in the web browser directly)

**Comment by keithmanville** [comment]
## Features and functionality

- Capability for interacting and viewing data files within the web UI
- Standardized format for storing and structuring different classes of dataset
- Defined metadata schema for tracking information about the data files and collective datasets
- Search support based on the metadata information
- Capability to augment datasets with artifacts produced during experiments, such as adversarial images
- Capability to automatically download popular/common datasets, such as MNIST, ImageNet, etc.

## Definition of Done

- [ ] Identify if any other features are needed for this endpoint
- [ ] The features and functionality are all captured in separate issue cards with their own definitions of done
